File: solutions/ruby/luhn/1/luhn.rb
module Luhn
  def self.valid? n
    return false if n.scan(/[0-9]/).length <= 1 || n.downcase.scan(/[a-z#$%-]/).length > 0
    n = n.scan(/[0-9]/)
    i = n.length - 1
    sum = 0
    mod = false
    
    while i >= 0
      d = n[i].to_i
      if mod
        d *= 2
        d -= 9 if d > 9
      end
        
      sum += d
      i -= 1
      mod = !mod
    end
    
    sum % 10 == 0
  end
end

File: solutions/c/luhn/1/luhn.c
#include "luhn.h"

bool luhn(const char *num)
{
    int sum = 0, count = 0, len = 0;
    
    while (num[len] != '\0') len++;

    for (int i = len - 1; i >= 0; i--) 
    {
        if (num[i] == ' ') continue;
        if (num[i] < '0' || num[i] > '9') return false;

        int value = num[i] - '0';

        if (count & 1) 
        {
            value <<= 1;
            if (value > 9) value -= 9;
        }

        sum += value;
        count++;
    }

    if (count < 2) return false;

    return (sum % 10 == 0);
}